src/moepkg/editor.nim

```diff
@@ -279,19 +251,259 @@ proc keyEnter*(bufStatus: var BufferStatus,
     windowNode.currentColumn = 0
     windowNode.expandedColumn = 0
 
+proc basicInsrtIndent(bufStatus: var BufferStatus,
+                      windowNode: WindowNode) =
+
+  let
+    currentLine = windowNode.currentLine
+    count = countRepeat(bufStatus.buffer[currentLine], Whitespace, 0)
+    indent = min(count, windowNode.currentColumn)
+
+    oldLine = bufStatus.buffer[currentLine + 1]
+  var newLine = bufStatus.buffer[currentLine + 1]
+
+  newLine &= repeat(' ', indent).toRunes
+  if oldLine != newLine:
+        bufStatus.buffer[currentLine + 1] = newLine
+
+proc insertIndetWhenPairOfParen(bufStatus: var BufferStatus,
+                                windowNode: WindowNode,
+                                autoIndent: bool,
+                                tabStop: int) =
+  let
+    currentLine = windowNode.currentLine
+    line = bufStatus.buffer[currentLine]
+    count = countRepeat(line, Whitespace, 0) + tabStop
+    openParen = line[windowNode.currentColumn - 1]
+
+  let oldLine = bufStatus.buffer[windowNode.currentLine + 1]
+  var newLine = bufStatus.buffer[currentLine + 1]
+  newLine &= repeat(' ', count).toRunes
+  if oldLine != newLine:
+    bufStatus.buffer[currentLine + 1] = newLine
+
+  bufStatus.basicNewLine(windowNode, autoIndent, tabStop)
+
+  # Add the new line and move the close paren if finish the next line with the close paren.
+  let nextLine = bufStatus.buffer[currentLine + 1]
+  if isCloseParen(nextLine[^1]) and
+     isCorrespondingParen(openParen, nextLine[^1]):
+    let closeParen = nextLine[^1]
+    # Delete the close paren in the nextLine
+    block:
+      let oldLine = nextLine
+      var newLine = nextLine
+      newLine = oldLine[0 .. newLine.high - 1]
+      if oldLine != newLine:
+        bufStatus.buffer[currentLine + 1] = newLine
+    # Add the close paren in the buffer[nextLine + 1]
+    block:
+      let count = countRepeat(line, Whitespace, 0)
+      var newLine = repeat(' ', count).toRunes & closeParen
+      bufStatus.buffer.insert(newLine, windowNode.currentLine + 1)
+
+proc insertIndentInNim(bufStatus: var BufferStatus,
+                       windowNode: WindowNode,
+                       autoIndent: bool,
+                       tabStop: int) =
+
+  let
+    currentLine = windowNode.currentLine
+    currentColumn = windowNode.currentColumn
+    line = bufStatus.buffer[currentLine]
+
+  if line.len > 0:
+    # Auto indent if the current line are "var", "let", "const".
+    # And, if finish the current line with ':', "object"
+    if (currentColumn == line.len) and
+       (line.splitWhitespace == @[ru "var"] or
+       line.splitWhitespace == @[ru "let"] or
+       line.splitWhitespace == @[ru "const"] or
+       (line.len > 6 and line[line.len - 6 .. ^1] == ru "object") or
```

Comment:
Why not:
`line.splitWhitespace[^1] == (ru "object") or`

Reply:
You are right. This should be ```line.splitWhitespace[^1] == (ru "object") or```

Reply:
Solved #1422

Reply:
Thanks!